### PR TITLE
[Bug Fix] Captions stuck at bottom 

### PR DESCRIPTION
**Pull request by carocao-msft**
# What
Updated captions processing logic to be closer to teams.
Only process the captions for the same speaker and ignore interceptors until the speaker stops talking (received a final caption)
Handle edge case with a 5000 ms time out, if we don't receive captions from the same speaker for over 5 seconds but the last received caption is still partial, turn partial into final and move on to the next speaker 

# Why
Solve the bug of caption stuck at bottom when not getting a final caption 

# How Tested
Tested with samples

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/calling-stateful-client/src/CallContext.ts [hunk 1]
```diff
@@ -976,24 +976,34 @@ export class CallContext {
   }
   /* @conditional-compile-remove(close-captions) */
   private processNewCaption(captions: CaptionsInfo[], newCaption: CaptionsInfo): void {
-    // going through current captions to find the last caption said by the same speaker, remove that caption if it's partial and replace with the new caption
-    for (let index = captions.length - 1; index >= 0; index--) {
-      const currentCaption = captions[index];
+    // time stamp when new caption comes in
+    newCaption.timestamp = new Date();
```
Comment (PorterNan) [inline packages/calling-stateful-client/src/CallContext.ts:980]: Do captions themselves have timestamp already?

File: packages/calling-stateful-client/src/CallContext.ts [hunk 2]
```diff
@@ -976,24 +976,34 @@ export class CallContext {
...
-        toFlatCommunicationIdentifier(currentCaption.speaker.identifier) ===
-          toFlatCommunicationIdentifier(newCaption.speaker.identifier)
+        toFlatCommunicationIdentifier(captions[captions.length - 1].speaker.identifier) ===
+        toFlatCommunicationIdentifier(newCaption.speaker.identifier)
       ) {
-        captions.splice(index, 1);
-        break;
+        captions[captions.length - 1] = newCaption;
+      }
+      //if different speaker, ignore the interjector until the current speaker finishes
```
Comment (PorterNan) [inline packages/calling-stateful-client/src/CallContext.ts:997]:
```suggestion
      // if different speaker, ignore the interjector until the current speaker finishes
```

File: packages/calling-stateful-client/src/CallContext.ts [hunk 3]
```diff
@@ -976,24 +976,34 @@ export class CallContext {
...
+      // edge case: if we dont receive the final caption from the current speaker for 5 secs, we turn the current speaker caption to final and push in the new interjector
+      else {
+        if (new Date().getTime() - captions[captions.length - 1].timestamp.getTime() > 5000) {
```
Comment (PorterNan) [inline packages/calling-stateful-client/src/CallContext.ts:1000]:
```suggestion
        if (new Date.now() - captions[captions.length - 1].timestamp.getTime() > 5000) {
```

**Review by PorterNan — APPROVED** [review]

File: packages/calling-stateful-client/src/CallContext.ts (hunk 1 above)
Comment (carocao-msft) [reply packages/calling-stateful-client/src/CallContext.ts:980]: Their timestamp is when the audio gets first inputted to the captions bot and it doesn't get updated. So we need to use our own time stamp here

**Review by edwardlee-msft — APPROVED** [review]

**Pull request merged**